relay-pattern/src/wildmatch.rs

```diff
@@ -0,0 +1,581 @@
+        //
+        // TODO: benchmark if allocation free is better/faster.
+        let needle = needle.as_case_converted_bytes();
+        let lower_haystack = haystack.to_lowercase();
+
+        memchr::arch::all::is_prefix(lower_haystack.as_bytes(), needle)
+            .then(|| recover_offset_len(haystack, 0, needle.len()).1)
+    }
+
+    #[inline(always)]
+    fn find(haystack: &str, needle: &Literal) -> Option<(usize, usize)> {
+        // TODO: implement manual lowercase which remembers if there were 'special' unicode
+        // conversion involved, if not, there is no recovery necessary.
+        // TODO: benchmark if a lut from offset -> original offset makes sense.
+        // TODO: benchmark allocation free and search with proper case insensitive search.
+        let needle = needle.as_case_converted_bytes();
+        let lower_haystack = haystack.to_lowercase();
+
+        let offset = memchr::memmem::find(lower_haystack.as_bytes(), needle)?;
+
+        // `offset` now points into the lowercase converted string, but this may not match the
+        // offset in the original string. Time to recover the index.
+        Some(recover_offset_len(haystack, offset, offset + needle.len()))
+    }
+
+    #[inline(always)]
+    fn ranges_match(c: char, negated: bool, ranges: &Ranges) -> bool {
+        let matches = exactly_one(c.to_lowercase()).is_some_and(|c| ranges.contains(c))
+            || exactly_one(c.to_uppercase()).is_some_and(|c| ranges.contains(c));
+        matches ^ negated
+    }
+}
+
+#[inline(always)]
+fn is_match_impl<T, M>(tokens: &T, haystack: &str) -> bool
+where
+    T: TokenIndex + ?Sized,
+    M: Matcher,
+{
+    // Remainder of the haystack which still needs to be matched.
+    let mut h_current = haystack;
+    // Saved haystack position for backtracking.
+    let mut h_revert = haystack;
+    // Revert index for `tokens`. In case of backtracking we backtrack to this index.
+    //
+    // If `t_revert` is zero, it means there is no currently saved backtracking position.
+    let mut t_revert = 0;
+    // The next token position which needs to be evaluted.
+    let mut t_next = 0;
+
+    macro_rules! advance {
+        ($len:expr) => {{
+            h_current = &h_current[$len..];
+            true
+        }};
+    }
+
+    // Empty glob never matches.
+    if tokens.is_empty() {
+        return false;
+    }
+
+    loop {
+        if t_next == tokens.len() {
+            return h_current.is_empty();
+        }
+
+        let token = &tokens[t_next];
+        t_next += 1;
+
+        let matched = match token {
+            Token::Literal(literal) => match M::is_prefix(h_current, literal) {
+                Some(n) => advance!(n),
+                None => false,
+            },
+            Token::Any(n) => {
+                advance!(match n_chars_to_bytes(*n, h_current) {
+                    Some(n) => n,
+                    None => return false,
+                });
+                true
```

Comment:
I think this would be slightly clearer:
```suggestion
                let Some(b) = match n_chars_to_bytes(*n, h_current) else {
                    return false;
                };
                advance!(b)
```
But it's also fine as it is.

Reply:
I added some docs to the branches which I think works better with the match.

Reply:
Yeah that's fair.